lab_4_summarization_textrank/README.md

```diff
@@ -0,0 +1,758 @@
+
+Рассмотрим пример. Пусть дана следующая последовательность токенов: `('в', 'лесу', 'родилась', 'ёлочка', 'в', 'лесу', 'она', 'росла')`
+В этой последовательности всего 6 разных строк: `'в'`, `'лесу'`, `'она'`, `'родилась'`, `'росла'`, `'ёлочка'`. 
+Присвоим каждой из этих строк по одному уникальному числу больше 1000. Допустим, соответствие получается такое:
+
+* `'в'` - `1009`
+* `'лесу'` - `1001`
+* `'она'` - `1005`
+* `'родилась'` - `1101`
+* `'росла'` - `1061`
+* `'ёлочка'` - `1041`
+
+Тогда атрибут `_word2id` должен выглядеть следующим образом: `{'в': 1009, 'лесу': 1001, 'она': 1005, 'родилась': 1101, 'росла': 1061, 'ёлочка': 1041}`. 
+В `_id2word`, соответственно, ключи и значения меняются местами: `{1009: 'в', 1001: 'лесу', 1005: 'она', 1101: 'родилась', 1061: 'росла', 1041: 'ёлочка'}`.
+
+Если вы все сделали правильно, то длина данных атрибутов должна совпадать.
+
+**Важно**: в рамках данной лабораторной работы метод должен **дополнять** атрибуты `_word2id` и `_id2word`, а не переписывать их. 
+Модифицируйте поведение метода, чтобы этот момент учитывался.
+
+Интерфейс:   
+```py
+def _learn_indices(self, tokens: tuple[str, ...]) -> None:
+    pass
+```
+
+#### Шаг 4.3 Кодирование предложений. Метод `encode_sentences`
+
+Этот метод кодирует предложения, то есть заменяет каждый токен на число. 
+
+Метод принимает последовательность предложений. 
+
+Метод добавляет к каждому предложению его закодированную версию.
+
+**Важно**: метод ничего не возвращает.
+
+**Важно №2**: на входящих предложениях необходимо использовать метод `_learn_indicies`.
+
+**Важно №3**: необходимо использовать методы `set_attribute` и `get_attribute` для взаимодействия с атрибутами экземпляров `Sentence`.  
+
+Интерфейс:   
+```py
+def encode_sentences(self, tokens: tuple[Sentence, ...]) -> None:
+    pass
+```  
+
+### Шаг 5. Продемонстрировать выделение и кодирование предложений. Выполнение Шагов 1-5 соответствует 6 баллам
+
+Выделите из текста, хранящегося в переменной `text`, предложения и закодируйте их. Это необходимо сделать в файле `start.py`.
+
+### Шаг 6. Представить текст в виде графа. Матрица схожести предложений
+
+В рамках [лабораторной работы №3](../lab_3_keywords_textrank/README.md) Вы реализовали класс `AdjacencyMatrixGraph`, позволяющий строить граф на основе совстречаемости токенов.
+Токены были вершинами, а связи между ними - рёбрами.
+
+В данной работе Вы модифицируете функционал этого класса, чтобы можно было работать с предложениями:
+* Вершинами будут предложения
+* Рёбрами будут значения похожести между предложения
+  * Расчёт похожести между предложениями будет рассмотрен в следующих шагах
+
+В этом шаге Вы будете реализовывать `SimilarityMatrix`.
+
+Матрица смежности представляет собой не что иное, как таблицу. В нашем случае незаполненная матрица выглядит вот так.
+
+|   |A   |B   |C   |D   |
+|---|---|---|---|---|
+|A   |   |   |   |   |
+|B   |   |   |   |   |
+|C   |   |   |   |   |
+|D   |   |   |   |   |
```

Comment:
and also it can be worth reminding that matrixx should contain only matrix without any service rows and columns